File: universal-adapter/parity/probe.ts
/**
 * TypeScript half of the cross-language parity runner.
 *
 * Do not run this by hand to answer a parity question. Run:
 *
 *     python universal-adapter/parity/run.py
 *
 * This program answers questions; it never judges the answers. It loads the
 * shared inputs (the fixture corpus and the case files in parity/cases/), asks
 * the TypeScript implementation what it makes of each one, and writes a flat map
 * of "group/key" -> "answer as a string" to stdout as JSON. probe.py answers the
 * same questions from the Python implementation, and run.py diffs the two maps.
 *
 * The two probes must stay structurally identical. If one of them ever starts
 * asking a question the other does not, run.py reports the key as missing rather
 * than quietly comparing fewer things than the README claims.
 *
 * Zero npm dependencies and no build step: Node 22+ executes this file and the
 * .ts sources it imports directly, which is the same claim the package itself
 * makes and would be embarrassing for its own parity runner to break.
 *
 * Usage: node probe.ts [path-to-pricing-table.json]
 * The pricing table argument carries Python's live gradebook.pricing table
 * across the language boundary. Without it the pricing answers are reported as
 * "unavailable:no_table_supplied" rather than skipped silently.
 */

import { readFileSync, existsSync } from "node:fs";
import { join, resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { canonicalize } from "../typescript/src/canonical.ts";
import { deriveDecisionId, contentDigest } from "../typescript/src/identity.ts";
import { validate } from "../typescript/src/validation.ts";
import { pricingTableId } from "../typescript/src/costing.ts";
import type { DecisionEvidenceBundle, PricingTable } from "../typescript/src/models.ts";

const HERE = dirname(fileURLToPath(import.meta.url));
const ADAPTER_ROOT = resolve(HERE, "..");
const FIXTURES = join(ADAPTER_ROOT, "fixtures");
const CASES = join(HERE, "cases");

function readJson(path: string): unknown {
  return JSON.parse(readFileSync(path, "utf8"));
}

/** The answer to every question is a string, including "it threw". */
function attempt(fn: () => string): string {
  try {
    return fn();
  } catch (error) {
    const name = error instanceof Error ? error.constructor.name : typeof error;
    return `error:${name}`;
  }
}

/**
 * Expand the $parity escapes the case files use for unspellable inputs. JSON has
 * no NaN, no infinities, and no comfortable way to write a 513 character string,
 * so the case files spell those as tagged objects and both probes expand them
 * identically before anything is measured.
 */
function decodeEscapes(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(decodeEscapes);
  if (value !== null && typeof value === "object") {
    const record = value as Record<string, unknown>;
    const tag = record["$parity"];
    if (typeof tag === "string" && Object.keys(record).length <= 3) {
      if (tag === "nan") return Number.NaN;
      if (tag === "inf") return Number.POSITIVE_INFINITY;
      if (tag === "-inf") return Number.NEGATIVE_INFINITY;
      if (tag === "repeat") return String(record["char"]).repeat(Number(record["count"]));
      throw new Error(`unknown $parity escape ${JSON.stringify(tag)}`);
    }
    const out: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(record)) out[key] = decodeEscapes(item);
    return out;
  }
  return value;
}

/** The JSON type name, which both languages must agree a document has. */
function jsonType(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return "array";
  switch (typeof value) {
    case "boolean":
      return "boolean";
    case "number":
      return "number";
    case "string":
      return "string";
    case "object":
      return "object";
    default:
      return `unrepresentable:${typeof value}`;
  }
}

interface CorpusCase {
  file: string;
  schema?: string;
  expect: "valid" | "invalid" | "not_validated";
  container?: "single" | "array";
}

/** A case is one document, or every element of one, as the index declares. */
function documents(testCase: CorpusCase, parsed: unknown): unknown[] {
  if (testCase.container === "array") {
    return Array.isArray(parsed) ? [...parsed] : [];
  }
  return [parsed];
}

function main(): void {
  const answers: Record<string, Record<string, string>> = {
    canonical: {},
    canonical_fixture: {},
    validation_corpus: {},
    validation_extra: {},
    decision_id: {},
    content_digest: {},
    pricing: {},
  };

  // 1. Canonical serialization over the adversarial value set.
  const canonicalCases = (readJson(join(CASES, "canonical.json")) as { cases: { id: string; value: unknown }[] }).cases;
  for (const testCase of canonicalCases) {
    const value = decodeEscapes(testCase.value);
    answers["canonical"]![testCase.id] = attempt(() => canonicalize(value));
  }

  // 2. The shared fixture corpus: validation verdicts, and the canonical form of
  //    every document in it.
  const index = readJson(join(FIXTURES, "index.json")) as { cases: CorpusCase[] };
  for (const testCase of index.cases) {
    const name = testCase.file;
    const path = join(FIXTURES, name);
    if (!existsSync(path)) {
      answers["validation_corpus"]![name] = "error:MissingFixture";
      continue;
    }
    const parsed = readJson(path);
    const docs = documents(testCase, parsed);

    if (testCase.expect === "not_validated") {
      // Raw harness input. Validating it against a protocol schema would be a
      // category error, so the only shared observation available is that both
      // languages parse it into the same shape.
      answers["validation_corpus"]![name] = `not_validated:${jsonType(parsed)}`;
    } else if (testCase.container === "array" && !Array.isArray(parsed)) {
      answers["validation_corpus"]![name] = "error:NotAnArray";
      continue;
    } else {
      docs.forEach((document, position) => {
        answers["validation_corpus"]![`${name}#${position}`] = attempt(() =>
          validate(testCase.schema ?? "", document).length === 0 ? "accepted" : "rejected",
        );
      });
    }

    docs.forEach((document, position) => {
      answers["canonical_fixture"]![`${name}#${position}`] = attempt(() => canonicalize(document));
    });

    // 3. Derived decision ids, for every VALID bundle document in the corpus.
    if (testCase.schema === "decision-evidence-bundle" && testCase.expect === "valid") {
      docs.forEach((document, position) => {
        const key = `${name}#${position}`;
        answers["decision_id"]![key] = attempt(() => deriveDecisionId(document as DecisionEvidenceBundle));
        answers["content_digest"]![key] = attempt(() => contentDigest(document as DecisionEvidenceBundle));
      });
    }
  }

  // 4. Extra validation inputs, at the type-system edges the corpus does not reach.
  const recordCases = (readJson(join(CASES, "records.json")) as { cases: { id: string; kind: string; record: unknown }[] })
    .cases;
  for (const testCase of recordCases) {
    const record = decodeEscapes(testCase.record);
    answers["validation_extra"]![testCase.id] = attempt(() =>
      validate(testCase.kind, record).length === 0 ? "accepted" : "rejected",
    );
  }

  // 5. The pricing table id. The table itself lives in Python's gradebook.pricing
  //    and cannot be imported here, so run.py exports it from probe.py and hands
  //    it over as a file. What is compared is the id the two implementations
  //    derive from the same table.
  const tablePath = process.argv[2];
  if (!tablePath) {
    answers["pricing"]!["pricing_table_id"] = "unavailable:no_table_supplied";
    answers["pricing"]!["pricing_table_canonical"] = "unavailable:no_table_supplied";
    answers["pricing"]!["pricing_table_model_count"] = "unavailable:no_table_supplied";
  } else {
    const table = readJson(tablePath) as PricingTable;
    answers["pricing"]!["pricing_table_id"] = attempt(() => pricingTableId(table));
    answers["pricing"]!["pricing_table_canonical"] = attempt(() => canonicalize(table));
    answers["pricing"]!["pricing_table_model_count"] = String(Object.keys(table.models ?? {}).length);
  }

  process.stdout.write(JSON.stringify({ implementation: "typescript", answers }));
}

main();

File: universal-adapter/parity/probe.py
"""Python half of the cross-language parity runner.

Do not run this by hand to answer a parity question. Run:

    python universal-adapter/parity/run.py

This program answers questions; it never judges the answers. It loads the shared
inputs (the fixture corpus and the case files in parity/cases/), asks the Python
implementation what it makes of each one, and writes a flat map of
"group/key" -> "answer as a string" to stdout as JSON. probe.ts answers the same
questions from the TypeScript implementation, and run.py diffs the two maps.

Keeping the judgement out of here is deliberate. A probe that decided for itself
whether an answer was acceptable could report agreement it had not observed, and
a parity runner that reports green on a check it did not really perform is worse
than no parity runner at all.

Every answer is a STRING, including failures, so that "both sides refused this
input" is a comparable observation rather than a crash. Failures are recorded as
"error:<ClassName>" and never as the error message: the two languages word their
messages differently on purpose, and comparing prose would produce noise that
buries a real divergence.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Callable

HERE = Path(__file__).resolve().parent
ADAPTER_ROOT = HERE.parent
FIXTURES = ADAPTER_ROOT / "fixtures"
CASES = HERE / "cases"

sys.path.insert(0, str(ADAPTER_ROOT / "python"))

from gradebook import pricing  # noqa: E402
from gradebook_adapter.costing import PRICING_TABLE_ID  # noqa: E402
from gradebook_adapter.models import (  # noqa: E402
    DecisionEvidenceBundle,
    bundle_content_digest,
    canonical_json,
    derive_decision_id,
)
from gradebook_adapter.validation import validate_record  # noqa: E402


def read_json(path: Path) -> Any:
    # Explicit utf-8. Python's default encoding is platform dependent and the
    # corpus carries non-ASCII, so an implicit decode would make this probe's
    # answers depend on which machine it ran on.
    return json.loads(path.read_text(encoding="utf-8"))


def attempt(fn: Callable[[], str]) -> str:
    try:
        return fn()
    except Exception as exc:  # noqa: BLE001 - the class name IS the answer
        return f"error:{type(exc).__name__}"


def decode_escapes(value: Any) -> Any:
    """Expand the $parity escapes the case files use for unspellable inputs.

    JSON has no NaN, no infinities, and no comfortable way to write a 513
    character string, so the case files spell those as tagged objects and both
    probes expand them identically before anything is measured.
    """
    if isinstance(value, list):
        return [decode_escapes(item) for item in value]
    if isinstance(value, dict):
        tag = value.get("$parity")
        if isinstance(tag, str) and len(value) <= 3:
            if tag == "nan":
                return float("nan")
            if tag == "inf":
                return float("inf")
            if tag == "-inf":
                return float("-inf")
            if tag == "repeat":
                return str(value["char"]) * int(value["count"])
            raise ValueError(f"unknown $parity escape {tag!r}")
        return {key: decode_escapes(item) for key, item in value.items()}
    return value


def json_type(value: Any) -> str:
    """The JSON type name, which both languages must agree a document has."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return f"unrepresentable:{type(value).__name__}"


def documents(case: dict[str, Any], parsed: Any) -> list[Any]:
    """A case is one document, or every element of one, as the index declares."""
    if case.get("container") == "array":
        if not isinstance(parsed, list):
            return []
        return list(parsed)
    return [parsed]


def main() -> int:
    answers: dict[str, dict[str, str]] = {
        "canonical": {},
        "canonical_fixture": {},
        "validation_corpus": {},
        "validation_extra": {},
        "decision_id": {},
        "content_digest": {},
        "pricing": {},
    }

    # 1. Canonical serialization over the adversarial value set.
    for case in read_json(CASES / "canonical.json")["cases"]:
        value = decode_escapes(case["value"])
        answers["canonical"][case["id"]] = attempt(lambda v=value: canonical_json(v))

    # 2. The shared fixture corpus: validation verdicts, and the canonical form
    #    of every document in it. The corpus is real protocol data, so it is the
    #    most representative serialization input available.
    index = read_json(FIXTURES / "index.json")
    for case in index["cases"]:
        name = case["file"]
        path = FIXTURES / name
        if not path.exists():
            answers["validation_corpus"][name] = "error:MissingFixture"
            continue
        parsed = read_json(path)
        docs = documents(case, parsed)

        if case["expect"] == "not_validated":
            # Raw harness input. Validating it against a protocol schema would
            # be a category error, so the only shared observation available is
            # that both languages parse it into the same shape.
            answers["validation_corpus"][name] = f"not_validated:{json_type(parsed)}"
        else:
            if case.get("container") == "array" and not isinstance(parsed, list):
                answers["validation_corpus"][name] = "error:NotAnArray"
                continue
            for position, document in enumerate(docs):
                key = f"{name}#{position}"
                verdict = attempt(
                    lambda d=document: "accepted"
                    if not validate_record(d, case["schema"])
                    else "rejected"
                )
                answers["validation_corpus"][key] = verdict

        for position, document in enumerate(docs):
            key = f"{name}#{position}"
            answers["canonical_fixture"][key] = attempt(lambda d=document: canonical_json(d))

        # 3. Derived decision ids, for every VALID bundle document in the corpus.
        #    Invalid bundles are excluded on purpose: see the README section
        #    "What this does not prove".
        if case["schema"] == "decision-evidence-bundle" and case["expect"] == "valid":
            for position, document in enumerate(docs):
                key = f"{name}#{position}"
                answers["decision_id"][key] = attempt(
                    lambda d=document: derive_decision_id(DecisionEvidenceBundle.from_dict(d))
                )
                answers["content_digest"][key] = attempt(
                    lambda d=document: bundle_content_digest(DecisionEvidenceBundle.from_dict(d))
                )

    # 4. Extra validation inputs, at the type-system edges the corpus does not
    #    reach.
    for case in read_json(CASES / "records.json")["cases"]:
        record = decode_escapes(case["record"])
        answers["validation_extra"][case["id"]] = attempt(
            lambda r=record: "accepted" if not validate_record(r, case["kind"]) else "rejected"
        )

    # 5. The pricing table id, over the real table in gradebook.pricing.
    #    TypeScript cannot import a Python module, so the table itself is
    #    exported here and handed to probe.ts by run.py. What is compared is the
    #    id the two implementations derive from the same table.
    table = {
        "currency": "usd",
        "models": {
            slug: {
                "input_per_1m_usd": rate.input_per_mtok,
                "output_per_1m_usd": rate.output_per_mtok,
            }
            for slug, rate in pricing.PRICES.items()
        },
    }
    answers["pricing"]["pricing_table_id"] = PRICING_TABLE_ID
    answers["pricing"]["pricing_table_canonical"] = attempt(lambda: canonical_json(table))
    answers["pricing"]["pricing_table_model_count"] = str(len(table["models"]))

    sys.stdout.write(
        json.dumps(
            {"implementation": "python", "answers": answers, "pricing_table": table},
            ensure_ascii=True,
            sort_keys=True,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
